CsvTextFieldParser/CsvTextFieldFormatter.cs

```diff
@@ -0,0 +1,293 @@
+        /// </summary>
+        public CsvTextFieldFormatter(Stream stream, Encoding encoding)
+            : this(new StreamWriter(stream, encoding)) { }
+
+        /// <summary>
+        /// Constructs a formatter from the specified output stream with the specified encoding, and optionally leaves the stream open.
+        /// </summary>
+        public CsvTextFieldFormatter(Stream stream, Encoding encoding, bool leaveOpen)
+            : this(new StreamWriter(stream, encoding, 1024, leaveOpen)) { }
+
+        /// <summary>
+        /// Constructs a formatter from the specified output file path.
+        /// </summary>
+        public CsvTextFieldFormatter(string path)
+            : this(new StreamWriter(path)) { }
+
+        /// <summary>
+        /// Constructs a formatter from the specified output file path with the specified encoding.
+        /// If the file exists, it can be either overwritten or appended to.
+        /// If the file does not exist, this constructor creates a new file.
+        /// </summary>
+        public CsvTextFieldFormatter(string path, bool append)
+            : this(new StreamWriter(path, append)) { }
+
+        /// <summary>
+        /// Constructs a formatter from the specified output file path with the specified encoding.
+        /// If the file exists, it can be either overwritten or appended to.
+        /// If the file does not exist, this constructor creates a new file.
+        /// </summary>
+        public CsvTextFieldFormatter(string path, bool append, Encoding encoding)
+            : this(new StreamWriter(path, append, encoding)) { }
+
+        /// <summary>
+        /// Constructs a formatter from the specified input text writer.
+        /// </summary>
+        public CsvTextFieldFormatter(TextWriter writer)
+        {
+            if (writer == null) throw new ArgumentNullException(nameof(writer));
+            this.writer = writer;
+        }
+
+        /// <summary>
+        /// Writes the fields to a new line.
+        /// </summary>
+        /// <param name="fields">An array of strings that contains field values for the current line.</param>
+        public void WriteFields(IEnumerable<string> fields)
+        {
+            if (writer == null) throw new ObjectDisposedException(nameof(writer));
+            if (fields == null) throw new ArgumentNullException(nameof(fields));
+
+            using (var enumerator = fields.GetEnumerator())
+            {
+                if (enumerator.MoveNext())
+                {
+                    WriteField(enumerator.Current);
+                    while (enumerator.MoveNext())
+                    {
+                        writer.Write(delimiterChar);
+                        WriteField(enumerator.Current);
+                    }
+                }
+            }
+            WriteEol();
+        }
+
+        private void WriteField(string field)
+        {
+            var isQuoteNecessary = field.IndexOfAny(SpecialChars) >= 0;

```

Comment:
as I suggested in the comment:
        /// <summary>
        /// Some cvs consumers trim leading and trailing white space from values unless they are escaped.
        /// </summary>
        public bool EscapeWhiteSpace { get; set; }

and

var isQuoteNecessary = field.IndexOfAny(SpecialChars) >= 0 ||
 (EscapeWhiteSpace && field.Length > 0 && (char.IsWhiteSpace(field[0]) || char.IsWhiteSpace(field[field.Length - 1])));




/// <summary>
/// Some cvs consumers trim leading and trailing white space from values unless they are escaped
/// </summary> (EscapeWhiteSpace && str != null && str.Length > 0 && (char.IsWhiteSpace(str[0]) || char.IsWhiteSpace(str[str.Length - 1])));
        public bool EscapeWhiteSpace { get; set; }

var isQuoteNecessary = field.IndexOfAny(SpecialChars) >= 0 ||
 (EscapeWhiteSpace && field .Length > 0 && (char.IsWhiteSpace(field[0]) || char.IsWhiteSpace(field[field.Length - 1])));

Reply:
I've been going back and forth on this in my head for a few days, and I think I've landed on preferring an option to always quote over an option that's just about whitespace handling.

My thoughts on why:

The parser in this library matches the behavior of Microsoft's VisualBasic TextFieldParser, and that parser doesn't consider quotes at all when it comes to whitespace trimming.  Depending on the TrimWhiteSpace option, it either always preserves whitespace (even in unquoted fields) or always trims whitespace (even if the field is enclosed in quotes).

So if the goal of the formatter is to act as a companion to the parser, I don't think an option to quote for leading or trailing whitespace is a good fit.

However, I do think an option to always quote would be reasonable.  The parser already has a "HasFieldsEnclosedInQuotes" option, so it already has a concept of a CSV file never using quotes.  A corresponding "ForceFieldsEnclosedinQuotes" option would cover the cases where you want to use the formatter with other parsers that handle leading/trailing whitespace differently, and might even handle other situations where you'd want to require quotes (like maybe a parser collapses whitespace in the middle of a field if it's unquoted).

Reply:
I agree. The white space logic is too specific.